library/dd-sdk-android-session-replay/src/main/kotlin/com/datadog/android/sessionreplay/internal/async/BlockingQueueHandler.kt

```diff
@@ -0,0 +1,135 @@
+
+        val blockingQueueItem = BlockingQueueItem(
+            timestamp = rumContextData.timestamp,
+            prevRumContext = rumContextData.prevRumContext,
+            newRumContext = rumContextData.newRumContext,
+            systemInformation = systemInformation
+        )
+
+        @Suppress("SwallowedException", "TooGenericExceptionCaught")
+        try {
+            workQueue.offer(blockingQueueItem)
+        } catch (e: IllegalArgumentException) {
+            // TODO: REPLAY-1364 Add logs here once the sdkLogger is added
+        } catch (e: ClassCastException) {
+            // in theory will never happen but we'll log it
+            // TODO: REPLAY-1364 Add logs here once the sdkLogger is added
+        } catch (e: NullPointerException) {
+            // in theory will never happen but we'll log it
+            // TODO: REPLAY-1364 Add logs here once the sdkLogger is added
+        }
+
+        return blockingQueueItem
+    }
+
+    @MainThread
+    internal fun update() {
+        // currentTime needs to be obtained on the uithread
+        val currentTime = timeProvider.getDeviceTimestamp()
+
+        @Suppress("SwallowedException", "TooGenericExceptionCaught")
+        try {
+            executorService.execute {
+                wakeUpProcessor(currentTime)
+            }
+        } catch (e: RejectedExecutionException) {
+            // TODO: REPLAY-1364 Add logs here once the sdkLogger is added
+        } catch (e: NullPointerException) {
+            // in theory will never happen but we'll log it
+            // TODO: REPLAY-1364 Add logs here once the sdkLogger is added
+        }
+    }
+
+    // end region
+
+    // region private
+
+    @WorkerThread
+    private fun wakeUpProcessor(currentTime: Long) {
+        removeInvalidItemsFromHead(currentTime)
+
+        while (workQueue.isNotEmpty()) {
+            val blockingQueueItem = workQueue.peek()
+
+            if (blockingQueueItem == null || blockingQueueItem.isNotReady()) {
+                break
+            }
+
+            workQueue.poll()
+            processor.processScreenSnapshots(
+                newContext = blockingQueueItem.newRumContext,
+                prevContext = blockingQueueItem.prevRumContext,
+                timestamp = blockingQueueItem.timestamp,
+                nodes = blockingQueueItem.nodes,
+                systemInformation = blockingQueueItem.systemInformation
+            )
+        }
+    }
+
+    private fun removeInvalidItemsFromHead(currentTime: Long) {
+        while (workQueue.isNotEmpty()) {
+            val blockingQueueItem = workQueue.peek()
+
+            if (shouldRemoveItem(blockingQueueItem, currentTime)) {
+                break
+            }
+
+            workQueue.poll()
+        }
+    }
+
+    private fun shouldRemoveItem(blockingQueueItem: BlockingQueueItem?, currentTime: Long) =
+        blockingQueueItem == null ||
+            (
+                blockingQueueItem.isValid() &&
```

Comment:
shouldn't be here `isNotValid && isTooOld`

Reply:
originally there was one loop that did exactly that, but I had issues with detekt and tried to split it up. I'll do some refactor and commenting

Reply:
We should always favour optimisation to `detekt`. Detekt rules can be suppressed if they alter optimisation.